core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -119,6 +119,8 @@ class ReplicaFetcherThread(name: String,
 
       if (fetchOffset != replica.logEndOffset.messageOffset)
         throw new RuntimeException("Offset mismatch for partition %s: fetched offset = %d, log end offset = %d.".format(topicAndPartition, fetchOffset, replica.logEndOffset.messageOffset))
+      if (!messageSet.isEmpty && messageSet.head.offset < replica.logEndOffset.messageOffset)
+        throw new RuntimeException("Offset error for partition %s: first offset in message set = %d, must not be less than log end offset = %d.".format(topicAndPartition, messageSet.head.offset, replica.logEndOffset.messageOffset))
```

Comment:
@imandhan : Instead of adding the checking here, it's probably better to do that in Log.append(). In the path where assignOffsets is false, we can just require that appendInfo.firstOffset >= nextOffsetMetadata.messageOffset. This also makes testing easier since it can be done on an instance of Log.